### PR TITLE
Refactor release upload API.

### PR DESCRIPTION
**Pull request by bloodearnest**
This is a refactoring of the Release and ReleaseFile models, and their
upload API.  Previously, a Release and its ReleaseFiles were created
atomically by single zip upload. This change separates creation of
a `Release` from uploading its `ReleaseFile`s, as well as tightens
various model contraints.

A user can now create a Release, and a different user can upload files
to that Release. This is to accomodate the intended workflow of
reviewing releases, and also the potential to upload some but not all
files requested.

In the process, some things have been simplified:

 - hashing and deduplication is now per file, rather than per zip, which
   was bugged.
 - removed `backend_user` and just use created_by, as we have a solution
   for that now on the TPP backend.
 - got rid of manifest uploading as it was not being used anywhere.
 - the API urls were simplified as a result and are more RESTish:
   - A single Workspace Releases API url, which on POST will create
     a release, and on GET will list workspace files.
   - A single RequestFile API url, which on PUT will upload and on GET
     will download.
   - A release index url, which GET will list files in release

Some things got more complex:

 - RequestFiles need to be uploading one by one, and validated that
   their filename, contents, and backend match the corresponding Release

Various other related changes:

 - Switch to UUID for Release.id
 - File listing API response changed from list of filenames:
    - top level dict with "files" key for security
    - files list of dicts for easier sorting
 - switched to sha256 to avoid md5 security silliness
 - fixed a ReleaseFile.get_absolute_path bug

What is not done in this PR, but will come later:

 - rewrite release management command to use new API
 - add permissions checks to uploading files
 - changing Release.status

### CODE REVIEW

File: jobserver/models/outputs.py [hunk 1]
```diff
@@ -95,7 +104,7 @@ def get_absolute_url(self):
                 "project_slug": self.release.workspace.project.slug,
                 "workspace_slug": self.release.workspace.name,
                 "pk": self.release.id,
-                "path": self.path,
+                "path": self.name,
```
Comment (bloodearnest) [inline jobserver/models/outputs.py:107]: This was a driveby bugfix - `self.path` is the path to the file on disk (relative to `RELEASES_STORAGE`), and shouldn't be part of a url.

`self.name` is the users name for the file (e.g. "output/file1.txt"), and is the part that should be used in the url

File: tests/jobserver/test_releases.py [hunk 2]
```diff
@@ -20,163 +25,89 @@ def raise_error(**kwargs):
...
-    assert release.backend_user == "user"
-    assert release.files.count() == 1
-    assert release.files.first().name == "file.txt"
-    assert (
-        release.files.first().path
-        == f"{workspace.name}/releases/{upload.hash}/file.txt"
-    )
-    assert release.manifest == {"workspace": "workspace", "repo": "repo"}
+    user = UserFactory()
+    releases.create_release(workspace, backend, user, {"file1.txt": "hash"})
```
Comment (ghickman) [inline tests/jobserver/test_releases.py:32]: Should there be some asserts after this?

File: jobserver/models/outputs.py (hunk 1 above)
Comment (bloodearnest) [reply jobserver/models/outputs.py:107]: As an aside, this is why I like hardcoding url expectations in tests, rather than asserting the output matchs a reverse lookup. That tests would have have passed, but the url is wrong.

Hardcoding url means we are encoding our url model outside of just urls.py, which is good to catch these errors. It does mean that when you change urls, there's some updating to do, but I think it comes on the right side of cost/benefit for me.

File: tests/jobserver/test_releases.py [hunk 3]
```diff
@@ -20,163 +25,89 @@ def raise_error(**kwargs):
...
+            release.backend,
+            release.created_by,
+            uploads[0].stream,
+            uploads[0].filename,
+        )
 
-    release_hash, release_files = hash_files(dirpath)
-    assert "metadata/manifest.json" not in release_files
-    assert release_hash == "6c52ca16d696574e6ab5ece283eb3f3d"
+    # check the fileextracted files have been deleted due to the error
```
Comment (ghickman) [inline tests/jobserver/test_releases.py:92]: s/fileextracted/extracted/ ?

File: tests/jobserver/api/test_releases.py [hunk 4]
```diff
@@ -82,140 +82,196 @@ def test_releasenotificationapicreate_with_failed_slack_update(
...
+    backend1 = BackendFactory(name="backend1")
+    backend2 = BackendFactory(name="backend2")
+    user = UserFactory()
     client = APIClient()
+    client.force_authenticate(user=user)
+    ProjectMembershipFactory(
+        user=user, project=workspace.project, roles=[ProjectCollaborator]
+    )
+
+    index_url = f"/api/v2/releases/workspace/{workspace.name}"
```
Comment (ghickman) [inline tests/jobserver/api/test_releases.py:221]: Can we keep the URL that's under test next to where it's used?  (Don't mind if that's inline to the `.get()` or just before it!)

File: tests/jobserver/api/test_releases.py [hunk 5]
```diff
@@ -224,10 +280,10 @@ def test_release_index_api_logged_in_no_permission():
 
 @pytest.mark.django_db
 def test_release_index_api_have_permission():
-    release = ReleaseFactory()
+    release = ReleaseFactory(ReleaseUploadsFactory(["file.txt"]))
     user = UserFactory()
     client = APIClient()
-    index_url = f"/api/v2/releases/{release.id}"
+    index_url = f"/api/v2/releases/release/{release.id}"
```
Comment (ghickman) [inline tests/jobserver/api/test_releases.py:286]: Same here.

**Review by ghickman — CHANGES_REQUESTED** [review]
This looks great!  Those layers for validating permissions and the new factories have worked out really well 👍

Bunch of small changes around naming, readability, etc

File: jobserver/api/releases.py [hunk 6]
```diff
@@ -43,44 +44,139 @@ def perform_create(self, serializer):
...
-class ReleaseUploadAPI(APIView):
+def validate_upload_access(request, workspace):
+    """Validate the request can upload releases for this workspace.
+
+    This validation uses backend authentication to authenticate the user, but
+    then checks that user has the correct permissons."""
+    # authenticate and get backend
+    backend = get_backend_from_token(request.headers.get("Authorization"))
+
+    # We trust the username supplied by the backend because it is authenticated.
```
Comment (evansd) [inline jobserver/api/releases.py:55]: Drive by comment but I found this line confusing because you can read "it" as referring to the username and then I wondered how that was authenticated.

Maybe a clearer way of putting it would be:
```py
# The request is from an authenticated backend so we trust it to supply arbitrary usernames
```

File: jobserver/api/releases.py [hunk 7]
```diff
@@ -43,44 +44,139 @@ def perform_create(self, serializer):
...
+            raise NotFound
+
+        backend, user = validate_upload_access(request, release.workspace)
+
+        # ensure this is being released from the same backend as the Release
+        # was created from
+        if release.backend != backend:
+            raise ValidationError(
+                {
+                    "detail": f"Release is from backend {release.backend.name} not {backend.name}"
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply jobserver/api/releases.py:156]: I don't think so in this case. It's an API error, intended for developers, so `backned.name` is what we want here I think?

File: jobserver/api/releases.py [hunk 8]
```diff
@@ -43,44 +44,139 @@ def perform_create(self, serializer):
...
+                }
+            )
 
         if "file" not in request.data:
             raise ValidationError({"detail": "No data uploaded"})
 
-        upload = request.data["file"]
-        release, created = handle_release(
-            workspace, backend, backend_user, release_hash, upload
+        releases.handle_release_upload(
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply jobserver/api/releases.py:163]: `handle_file_upload` ?

File: jobserver/models/outputs.py [hunk 9]
```diff
@@ -15,9 +14,12 @@ def absolute_file_path(path):
 class Release(models.Model):
     """A set of reviewed and redacted outputs from a Workspace"""
 
-    # No value in the default Autoid as we are using a content-addressable hash
-    # as it. Additionaly it avoids enumeration attacks.
-    id = models.TextField(primary_key=True)
+    class ReleaseStatus(models.TextChoices):
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply jobserver/models/outputs.py:17]: :+1:

File: jobserver/models/outputs.py [hunk 10]
```diff
@@ -79,10 +77,19 @@ class ReleaseFile(models.Model):
...
+        "User",
+        on_delete=models.PROTECT,
+        related_name="released_files",
+    )
     # name is path from the POV of the researcher, e.g "outputs/file1.txt"
     name = models.TextField()
     # path is from the POV of the system e.g. "workspace/release/outputs/file1.txt"
     path = models.TextField()
+    # the sha256 hash of the file (temporary default to enable migration).
+    filehash = models.TextField()
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply jobserver/models/outputs.py:94]: old comment from previous migration attempt, will remove

File: jobserver/releases.py [hunk 11]
```diff
@@ -1,144 +1,75 @@
...
+    """Validate and save an uploaded file to disk and database.
+
+    Does basic detection of re-uploads of the same file, to avoid duplication.
+    """
+    # This reads the whole file into memory, but should not be a problem.
+    # TODO: enforce a max upload size?
+    data = upload.read()
+    calculated_hash = hashlib.sha256(data).hexdigest()
+
+    # idempotentancy
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply jobserver/releases.py:33]: Can't change it, it's idempotent

[the remark this answers is not in the document]
Comment (bloodearnest) [reply jobserver/releases.py:33]: :stuck_out_tongue:

File: jobserver/models/outputs.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bloodearnest) [reply jobserver/models/outputs.py:17]: Surely `ReleaseStatii` !

File: tests/jobserver/test_releases.py (hunk 2 above)
Comment (bloodearnest) [reply tests/jobserver/test_releases.py:32]: do'h, although, there's not much to assert it was basically checking the db object was created without raising.

File: tests/jobserver/api/test_releases.py (hunk 5 above)
Comment (bloodearnest) [reply tests/jobserver/api/test_releases.py:286]: Yeah, these are left over from when I had all the permissions checks in the same test, rather than split into individual tests.

File: jobserver/api/releases.py (hunk 6 above)
Comment (bloodearnest) [reply jobserver/api/releases.py]: Fixed

File: jobserver/api/releases.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ghickman) [reply jobserver/api/releases.py]: Sounds good!

File: jobserver/releases.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ghickman) [reply jobserver/releases.py]: Simon.

File: jobserver/api/releases.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bloodearnest) [reply jobserver/api/releases.py]: done

**Review by ghickman — APPROVED** [review]
🎉
